BSP/STM32F413/Src/BSP_CAN.c

```diff
@@ -0,0 +1,374 @@
+const CAN_HandleTypeDef *CarCAN = &CarCAN_;
+const CAN_HandleTypeDef *LocalCAN = &LocalCAN_;
+
+/* communicators keep track of queues to use for communication */
+typedef struct CAN_Listener
+{
+    uint32_t id;            /* can id */
+    QueueHandle_t *queue;   /* queue to wait on */
+} CAN_Listener_t;
+
+#define NUM_LISTENERS 16
+
+/* array of listeners */
+static CAN_Listener_t CarCAN_Listeners[NUM_LISTENERS] = {0};
+static CAN_Listener_t LocalCAN_Listeners[NUM_LISTENERS] = {0};
+
+/* transmit queues */
+QueueHandle_t CarCAN_TxQueue = (QueueHandle_t){0};
+QueueHandle_t LocalCAN_TxQueue = (QueueHandle_t){0};
+
+#define TX_QUEUE_LEN 32
+#define TX_QUEUE_ITEM_SIZE sizeof(CAN_TxPayload_t)
+
+/* queue storage */
+uint8_t CarCAN_Payloads[TX_QUEUE_LEN * TX_QUEUE_ITEM_SIZE] = {0};
+uint8_t LocalCAN_Payloads[TX_QUEUE_LEN * TX_QUEUE_ITEM_SIZE] = {0};
+
+/* transmit queue buffers */
+StaticQueue_t CarCAN_QueueBuffer = (StaticQueue_t){0};
+StaticQueue_t LocalCAN_QueueBuffer = (StaticQueue_t){0};
+
+void CAN_CarCANInit(void)
+{
+    /* init transmit queue */
+    CarCAN_TxQueue = xQueueCreateStatic( 
+                        TX_QUEUE_LEN,
+                        TX_QUEUE_ITEM_SIZE,
+                        CarCAN_Payloads,
+                        &CarCAN_QueueBuffer);
+
+    /* init struct */
+    CarCAN_.Instance = CAN1;
+    CarCAN_.Init.Prescaler = 16;
+    CarCAN_.Init.Mode = CAN_MODE_NORMAL;
+    CarCAN_.Init.SyncJumpWidth = CAN_SJW_1TQ;
+    CarCAN_.Init.TimeSeg1 = CAN_BS1_1TQ;
+    CarCAN_.Init.TimeSeg2 = CAN_BS2_1TQ;
+    CarCAN_.Init.TimeTriggeredMode = DISABLE;
+    CarCAN_.Init.AutoBusOff = DISABLE;
+    CarCAN_.Init.AutoWakeUp = DISABLE;
+    CarCAN_.Init.AutoRetransmission = DISABLE;
+    CarCAN_.Init.ReceiveFifoLocked = DISABLE;
+    CarCAN_.Init.TransmitFifoPriority = DISABLE;
+
+    /* initialize can perph */
+    if (HAL_CAN_Init(&CarCAN_) != HAL_OK)
+    {
+        Error_Handler();
+    }
+
+    /* start can perph */
+    HAL_CAN_Start(&CarCAN_);
+}
+
+void CAN_LocalCANInit(void)
+{
+    /* init transmit queue */
+    LocalCAN_TxQueue = xQueueCreateStatic( 
+                            TX_QUEUE_LEN,
+                            TX_QUEUE_ITEM_SIZE,
+                            LocalCAN_Payloads,
+                            &LocalCAN_QueueBuffer);
+
+    /* init struct */
+    LocalCAN_.Instance = CAN3;
+    LocalCAN_.Init.Prescaler = 16;
+    LocalCAN_.Init.Mode = CAN_MODE_NORMAL;
+    LocalCAN_.Init.SyncJumpWidth = CAN_SJW_1TQ;
+    LocalCAN_.Init.TimeSeg1 = CAN_BS1_1TQ;
+    LocalCAN_.Init.TimeSeg2 = CAN_BS2_1TQ;
+    LocalCAN_.Init.TimeTriggeredMode = DISABLE;
+    LocalCAN_.Init.AutoBusOff = DISABLE;
+    LocalCAN_.Init.AutoWakeUp = DISABLE;
+    LocalCAN_.Init.AutoRetransmission = DISABLE;
+    LocalCAN_.Init.ReceiveFifoLocked = DISABLE;
+    LocalCAN_.Init.TransmitFifoPriority = DISABLE;
+
+    /* initialize can perph */
+    if (HAL_CAN_Init(&LocalCAN_) != HAL_OK)
+    {
+        Error_Handler();
+    }
```

Comment:
Need to enable interrupts. Look up / intellisense what your CAN equivalent of SPI1_IRQn is, it's all hardcoded into HAL

HAL_NVIC_SetPriority(SPI1_IRQn, 0, 0);
HAL_NVIC_EnableIRQ(SPI1_IRQn);